src/mission_companion.cpp

```diff
@@ -1563,3 +3685,672 @@ std::vector<item*> talk_function::loot_building(const tripoint site)
+        cooking_recipes.insert(COOK_SKILL_4.begin(), COOK_SKILL_4.end());
+    }
+    if( om_min_level( "faction_base_kitchen_6", om_cur ) || om_cur == "ALL" ){
+        cooking_recipes.insert(COOK_SKILL_5.begin(), COOK_SKILL_5.end());
+    }
+    if( om_min_level( "faction_base_kitchen_7", om_cur ) || om_cur == "ALL" ){
+        cooking_recipes.insert(COOK_SKILL_6.begin(), COOK_SKILL_6.end());
+    }
+    if( om_min_level( "faction_base_kitchen_8", om_cur ) || om_cur == "ALL" ){
+        cooking_recipes.insert(COOK_SKILL_7.begin(), COOK_SKILL_7.end());
+    }
+    return cooking_recipes;
+}
+
+int talk_function::camp_recipe_batch_max( const recipe making, inventory total_inv ){
+    int max_batch = 0;
+    int max_checks = 9;
+    int iter = 0;
+    while( iter < max_checks ) {
+        if( making.requirements().can_make_with_inventory( total_inv, max_batch + 1000 ) &&
+           camp_food_supply() > (max_batch + 1000) * time_to_food ( time_duration::from_turns( making.time / 100 ) ) ){
+            max_batch += 1000;
+        }
+        iter++;
+    }
+    iter = 0;
+    while( iter < max_checks ) {
+        if( making.requirements().can_make_with_inventory( total_inv, max_batch + 100 ) &&
+           camp_food_supply() > (max_batch + 100) * time_to_food ( time_duration::from_turns( making.time / 100 ) ) ){
+            max_batch += 100;
+        }
+        iter++;
+    }
+    iter = 0;
+    while( iter < max_checks ) {
+        if( making.requirements().can_make_with_inventory( total_inv, max_batch + 10 ) &&
+           camp_food_supply() > (max_batch + 10) * time_to_food ( time_duration::from_turns( making.time / 100 ) ) ){
+            max_batch += 10;
+        }
+        iter++;
+    }
+    iter = 0;
+    while( iter < max_checks ) {
+        if( making.requirements().can_make_with_inventory( total_inv, max_batch + 1 ) &&
+           camp_food_supply() > (max_batch + 1) * time_to_food ( time_duration::from_turns( making.time / 100 ) ) ){
+            max_batch += 1;
+        }
+        iter++;
+    }
+    return max_batch;
+}
+
+int talk_function::companion_skill_trainer( npc &comp, std::string skill_tested, time_duration time_worked, int difficulty ){
+    difficulty = ((difficulty <= 0) ? difficulty : 1);
+    float checks = to_minutes<int>( time_worked )/10.0;
+    int total = difficulty * checks;
+    int i = 0;
+    int y = 0;
+    if( skill_tested == "" || skill_tested == "gathering" ){
+        while( i < checks ){
+            y = rng( 0, 100 );
+            if (y < 80){
+                comp.practice( skill_survival, difficulty);
+            } else if (y < 85){
+                comp.practice( skill_traps, difficulty);
+            } else if (y < 90){
+                comp.practice( skill_fabrication, difficulty);
+            } else if (y < 95){
+                comp.practice( skill_archery, difficulty);
+            } else{
+                comp.practice( skill_melee, 5);
+            }
```

Comment:
`weighted_int_list` type was create for this exact purpose.
You use it like this:
```
weighted_int_list<skill_id> gathering_practice;
gathering_practice.add( skill_survival, 80 );
gathering_practice.add( skill_traps, 5 );
...
const skill_id &skill_to_practice = *gathering_practice.pick();
```

It's much better than manual `if/else` ladder with `rng( 0, 100 )`, because it guarantees that all rolled values have specific weights and those weights can be clearly seen.